File: leetcode/inorderTraversal/inorderTraversal.cpp
#include <vector>
#include <iostream>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
class Solution {
public:
    vector<int> res;
    void traversal(TreeNode* subNode){
        if(subNode == NULL)
            return;
        traversal(subNode->left);
        res.push_back(subNode->val);
        traversal(subNode->right);

    }
    vector<int> inorderTraversal(TreeNode* root) {
        res.clear();
        traversal(root);
        return res;
        
    }
};

File: leetcode/inorderTraversal/inorderTraversal.py
# Given a binary tree, return the inorder traversal of its nodes' values.

# Example:

# Input: [1,null,2,3]
#    1
#     \
#      2
#     /
#    3

# Output: [1,3,2]
# Follow up: Recursive solution is trivial, could you do it iteratively?

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.res = []

    def traversal(self, subNode):
        if subNode is None:
            return 
        self.traversal(subNode.left)
        self.res.append(subNode.val)
        self.traversal(subNode.right)

    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        self.res.clear()
        self.traversal(root)
        return self.res


if __name__ == '__main__':
    solver = Solution()

    node1 = TreeNode(1)
    node2 = TreeNode(2)
    node3 = TreeNode(3)
    node1.right = node2
    node2.left = node3
    root = node1

    print(solver.inorderTraversal(root))
